cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs

```diff
@@ -177,3 +215,158 @@ runGovernanceUpdateProposal (OutputFile upFile) eNo genVerKeyFiles upPprams mCos
+      , "Hint (1): Use '--json-metadata-detailed-schema' and '--metadata-json-file' "
+      , "from the build or build-raw commands.\n"
+      , "Hint (2): You can redirect the standard output of this command to a JSON "
+      , "file to capture metadata.\n\n"
+      , "Note: A serialized version of the poll suitable for sharing with "
+      , "participants has been generated at '" <> outPath <> "'."
+      ]
+
+runGovernanceAnswerPoll
+  :: FilePath
+  -> SigningKeyFile
+    -- ^ VRF or Ed25519 cold key
+  -> Maybe Word
+    -- ^ Answer index
+  -> ExceptT ShelleyGovernanceCmdError IO ()
+runGovernanceAnswerPoll pollFile skFile maybeChoice = do
+  poll <- firstExceptT ShelleyGovernanceCmdTextEnvReadError . newExceptT $
+    readFileTextEnvelope AsGovernancePoll pollFile
+
+  credentials <- readVRFOrColdSigningKeyFile skFile
+
+  choice <- case maybeChoice of
+    Nothing -> do
+      askInteractively poll
+    Just ix -> do
+      validateChoice poll ix
+      liftIO $ BSC.hPutStrLn stderr $ Text.encodeUtf8 $ Text.intercalate "\n"
+        [ govPollQuestion poll
+        , "→ " <> (govPollAnswers poll !! fromIntegral ix)
+        , ""
+        ]
+      pure ix
+
+  let pollAnswer = GovernancePollAnswer
+        { govAnsPoll = hashGovernancePoll poll
+        , govAnsChoice = choice
+        }
+  let witness = pollAnswer `signPollAnswerWith` credentials
+
+  let metadata =
+        mergeTransactionMetadata
+          ( \l r -> case (l, r) of
+              (TxMetaMap xs, TxMetaMap ys) -> TxMetaMap (xs <> ys)
+              _ -> error "unreachable"
+          )
+          (asTxMetadata pollAnswer)
+          (asTxMetadata witness)
+        & metadataToJson TxMetadataJsonDetailedSchema
+
+  liftIO $ do
+    BSC.hPutStrLn stderr $ mconcat
+      [ "Poll answer created successfully.\n"
+      , "Please submit a transaction using the resulting metadata.\n"
+      ]
+    BSC.hPutStrLn stdout (prettyPrintJSON metadata)
+    BSC.hPutStrLn stderr $ mconcat
+      [ "\n"
+      , "Hint (1): Use '--json-metadata-detailed-schema' and '--metadata-json-file' "
+      , "from the build or build-raw commands.\n"
+      , "Hint (2): You can redirect the standard output of this command to a JSON "
+      , "file to capture metadata."
+      ]
+ where
+  readVRFOrColdSigningKeyFile
+    :: SigningKeyFile
+    -> ExceptT
+         ShelleyGovernanceCmdError
+         IO
+         (Either (SignKeyVRF StandardCrypto) (SignKeyDSIGN StandardCrypto))
+  readVRFOrColdSigningKeyFile filepath = do
+    someSk <- firstExceptT ShelleyGovernanceCmdKeyReadError $
+      readSigningKeyFile filepath
+    case someSk of
+      AVrfSigningKey (VrfSigningKey sk) ->
+        pure (Left sk)
+      AStakePoolSigningKey (StakePoolSigningKey sk) ->
+        pure (Right sk)
+      _anythingElse ->
+        left $ ShelleyGovernanceCmdUnexpectedKeyType
+          [ textEnvelopeType (AsSigningKey AsVrfKey)
+          , textEnvelopeType (AsSigningKey AsStakePoolKey)
+          ]
```

Comment:
Note to self: Maybe I wonder if it's possible to report in the error what kind of signing key was actually found?

Reply:
It is, but it's a bit annoying because one needs to "unlift" the key type from a `SomeSigningKey`. So, feel free to do it.